### PR TITLE
feat: DB2 preprocessor statements highlighting and parsing 

### PR DESCRIPTION
**Pull request by jirimosinger**

### CODE REVIEW

File: utils/include/utils/string_operations.h [hunk 1]
```diff
@@ -21,7 +21,44 @@
...
+
+    const auto to_trim = s.find_first_not_of(' ');
+    if (to_trim == S::npos)
+    {
+        auto s_length = s.length();
+        s = {};
+        return s_length;
+    }
+
+    s.remove_prefix(to_trim);
```
Comment (slavek-kucera) [inline utils/include/utils/string_operations.h:38]: `S` can really only be std::string_view...

File: parser_library/src/diagnostic.cpp [hunk 2]
```diff
@@ -2348,18 +2348,33 @@ diagnostic_op diagnostic_op::error_DB004(const range& range)
...
-        std::string("DB2 preprocessor - continuation detected on SQL TYPE statement"),
+        std::string("DB2 preprocessor - continuation detected on 'SQL TYPE' statement"),
         range);
 }
 
+diagnostic_op diagnostic_op::warn_DB006(const range& range)
+{
+    return diagnostic_op(diagnostic_severity::warning,
+        "DB006",
+        std::string("DB2 preprocessor - requested 'SQL TYPE' not recognized (operand 'IS' either missing or split)"),
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/diagnostic.cpp:2368]: Well, I wanted to distinguish between missing operands and missing `IS` keyword to make `SQL TYPE` related diagnostics more granular. Also, `IS` doesn't have to be on the same line as `SQL TYPE`, IIRC.

So, when someone types `SQL TYPE operand operand` or when `IS` is split on two lines, the extension would complain just about the missing `IS`. 
And when someone types `SQL TYPE IS`, the extension would complain about missing operands.

It helped me a little when I was writing tests so I decided to leave it in.

Anyway, I can remove it if it still feels redundant to you.

File: parser_library/src/lexing/logical_line.h [hunk 3]
```diff
@@ -63,7 +63,9 @@ struct logical_line
     bool so_si_continuation;
     bool missing_next_line;
 
-    void clear()
+    virtual ~logical_line() noexcept = default;
+
+    virtual void clear() noexcept
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/lexing/logical_line.h:68]: Reworked

File: parser_library/src/processing/preprocessors/db2_preprocessor.cpp [hunk 4]
```diff
@@ -28,63 +32,310 @@
...
+#include "utils/string_operations.h"
+#include "utils/unicode_text.h"
 #include "workspaces/parse_lib_provider.h"
 
+namespace hlasm_plugin::parser_library::processing {
 namespace {
-constexpr std::string_view trim_right(std::string_view s)
+using utils::concat;
+
+struct db2_logical_line final : public lexing::logical_line
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/processing/preprocessors/db2_preprocessor.cpp:51]: Reworked by using 2 logical lines

File: parser_library/src/processing/preprocessors/db2_preprocessor.cpp [hunk 5]
```diff
@@ -28,63 +32,310 @@
...
+    lexing::finish_logical_line(out, opts);
 
-class db2_preprocessor final : public preprocessor // TODO Take DBCS into account
+    if (out.segments.empty())
+        return;
+
+    out.m_first_line = out.segments.front().code;
+    out.m_instruction_end = instruction_end;
+
+    std::stack<unsigned char> quotes;
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/processing/preprocessors/db2_preprocessor.cpp:192]: It does, indeed!

File: parser_library/src/processing/preprocessors/db2_preprocessor.cpp [hunk 6]
```diff
@@ -237,71 +488,123 @@ class db2_preprocessor final : public preprocessor // TODO Take DBCS into accoun
...
+            s.append("(?:[ ]|--)+(?:").append(*w_it++).append(")");
+
+        s.append(")([ ]|--)");
+        if (tolerate_no_space_at_end)
+            s.append("*");
+        else
+            s.append("+");
+        s.append("(.*)");
+
+        return std::regex(s);
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/processing/preprocessors/db2_preprocessor.cpp:509]: Reused parts made into static variables

File: parser_library/src/processing/preprocessors/db2_preprocessor.cpp [hunk 7]
```diff
@@ -466,127 +767,136 @@ class db2_preprocessor final : public preprocessor // TODO Take DBCS into accoun
...
-                    case 'O':
-                        if (!consume_words(operands, { "ROWID" }, true))
-                            break;
-                        add_ds_line(label, "", "H,CL40");
-                        return true;
+                    switch (*it++)
+                    {
+                        case 'E':
+                            if (!consume_words_advance_to_next(
+                                    it, it_e, { "SULT_SET_LOCATOR", "VARYING" }, false, true))
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/processing/preprocessors/db2_preprocessor.cpp]: Extracted to a separate function and the words are compared from the beginning.

File: parser_library/src/processing/preprocessors/db2_preprocessor.cpp [hunk 8]
```diff
@@ -820,12 +1150,65 @@ class db2_preprocessor final : public preprocessor // TODO Take DBCS into accoun
...
     }
 
+    void do_highlighting(const semantics::preprocessor_statement_si& stmt,
+        const lexing::logical_line& ll,
+        semantics::source_info_processor& src_proc,
+        size_t continue_column = 15) const override
+    {
+        preprocessor::do_highlighting(stmt, ll, src_proc, continue_column);
+
+        auto db2_ll = dynamic_cast<const db2_logical_line&>(ll);
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/processing/preprocessors/db2_preprocessor.cpp]: Reworked

File: parser_library/src/processing/preprocessors/db2_preprocessor.cpp [hunk 9]
```diff
@@ -820,12 +1150,65 @@ class db2_preprocessor final : public preprocessor // TODO Take DBCS into accoun
...
+                        range(position(lineno, operand_start_column), position(lineno, comment_start_column)),
+                        semantics::hl_scopes::operand));
+        }
+    }
+
+    db2_logical_line extract_nonempty_db2_logical_line(
+        size_t instruction_end, line_iterator& it, const line_iterator& end) const
+    {
+        db2_logical_line out;
+        out.clear();
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/processing/preprocessors/db2_preprocessor.cpp]: Code removed

File: utils/include/utils/string_operations.h (hunk 1 above)
Comment (jirimosinger) [reply utils/include/utils/string_operations.h]: Template removed and reworked related code in `low_language_processor.cpp`

File: utils/include/utils/string_operations.h [hunk 10]
```diff
@@ -21,7 +21,44 @@
...
+size_t trim_left(S& s)
+{
+    if (s.empty() || s.front() != ' ')
+        return 0;
+
+    const auto to_trim = s.find_first_not_of(' ');
+    if (to_trim == S::npos)
+    {
+        auto s_length = s.length();
+        s = {};
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply utils/include/utils/string_operations.h]: well, that's unfortunate.

File: parser_library/src/processing/preprocessors/db2_preprocessor.cpp [hunk 11]
```diff
@@ -12,79 +12,346 @@
...
+};
+
+template<typename It>
+class mini_parser
+{
+    void skip_to_matching_character(It& b, const It& e, char to_match) const
+    {
+        if (b == e)
+            return;
+
```
Comment (slavek-kucera) [inline parser_library/src/processing/preprocessors/db2_preprocessor.cpp:198]:
```suggestion
        const auto to_match = *b;
```

File: parser_library/src/lexing/logical_line.h [hunk 12]
```diff
@@ -132,19 +136,47 @@ struct logical_line
...
         friend bool operator!=(const const_iterator& a, const const_iterator& b) noexcept { return !(a == b); }
 
         bool same_line(const const_iterator& o) const noexcept
         {
             assert(m_logical_line == o.m_logical_line);
-            return m_segment == o.m_segment;
+            return m_segment_it == o.m_segment_it;
+        }
+
+        const_iterator transform_into(const logical_line& other_ll) const noexcept
```
Comment (slavek-kucera) [inline parser_library/src/lexing/logical_line.h:149]: I think I am starting to see what is the overall direction and use case here... so what about
```
std::pair<size_t, size_t> get_coordinates() const noexcept
{
    assert(m_col != std::string_view::const_iterator()); // valid only for non-end iterators
    return { m_segment->code_offset + std::distance(m_segment.code.begin(), m_col ), std::distance(m_logical_line->segments.begin(), m_segment)};
}
```
given how our range works, it should probably be code_offset_utf16 + utf16 distance (but I wouldn't bother with this at the moment).

Range between two iterators is then
```
range text_range(logical_line::const_iterator b, logical_line::const_iterator e) noexcept
{
    assert("b != ll.end()");
    const auto [bx, by] = b.get_coordinates();
    const position b_pos(bx, by);
    if (b == e) // empty range
        return range(b_pos, b_pos);
    else
        return range(b_pos, "std::prev(e).get_coordinates() + 1 column (really next character size in utf-16 )");
}
```

File: parser_library/src/processing/instruction_sets/low_language_processor.cpp [hunk 13]
```diff
@@ -89,50 +90,31 @@ bool low_language_processor::create_symbol(
...
-    std::string new_label;
+    const auto label_inserter = [&result, this](const std::string label, const range& r) {
+        if (std::string_view label_sv = label; utils::trim_right(label_sv) == label.length())
+            result.label.emplace(r);
+        else
+        {
+            auto ord_id = this->hlasm_ctx.ids().add(label_sv);
+            result.label.emplace(r, ord_symbol_string { ord_id, std::string(label_sv) });
+        }
+    };
```
Comment (slavek-kucera) [inline parser_library/src/processing/instruction_sets/low_language_processor.cpp:106]:
```suggestion
    const auto label_inserter = [&result, &ids = hlasm_ctx.ids()](std::string label, const range& r) {
        label.erase(label.find_last_not_of(' ') + 1);
        if (label.empty())
            result.label.emplace(r);
        else
        {
            auto ord_id = ids.add(label);
            result.label.emplace(r, ord_symbol_string { ord_id, std::move(label) });
        }
    };
```

File: parser_library/test/processing/db2_preprocessor_test.cpp [hunk 14]
```diff
@@ -404,7 +481,21 @@ TEST(db2_preprocessor, include_nonexistent)
...
+    mock_parse_lib_provider libs({
+        { "MEMBER", "" },
+    });
+    std::string input = " EXEC SQL INCLUDEMEMBER ";
+
+    analyzer a(input, analyzer_options { &libs, db2_preprocessor_options {} });
+    a.analyze();
+    a.collect_diags();
+
+    EXPECT_EQ(a.hlasm_ctx().get_visited_files().count(member_loc), 0);
```
Comment (slavek-kucera) [inline parser_library/test/processing/db2_preprocessor_test.cpp:498]: do we produce a diagnostic, or do we just let it fail in the opencode?

File: parser_library/src/processing/preprocessors/db2_preprocessor.cpp [hunk 15]
```diff
@@ -768,6 +1071,7 @@ class db2_preprocessor final : public preprocessor // TODO Take DBCS into accoun
     void generate_replacement(line_iterator it, line_iterator end, bool include_allowed)
     {
         bool skip_continuation = false;
+        auto ll = std::make_shared<lexing::logical_line>();
```
Comment (slavek-kucera) [inline parser_library/src/processing/preprocessors/db2_preprocessor.cpp:1074]: I don't understand why a new instance would need to be created for each line.
It seems perfectly reasonable to have reusable buffer cleared on each iteration and pass its reference into the db2_helper, or just have db2_helper with two logical_lines that could be also reused.

File: parser_library/test/processing/db2_preprocessor_test.cpp (hunk 14 above)
Comment (jirimosinger) [reply parser_library/test/processing/db2_preprocessor_test.cpp:498]: - It behaves like if it was any other ` EXEC SQL ` statement with non include operands like ` EXEC SQL op1 op2`

- It produces a bunch of errors coming from opencode

File: parser_library/src/lexing/logical_line.h (hunk 12 above)
Comment (jirimosinger) [reply parser_library/src/lexing/logical_line.h:149]: Seems to work perfectly! Nonetheless, where would you put `range text_range(...) noexcept` definition?

I have it in `db2_logical_line_helper` so far but maybe you've had more general location in mind like `range_provider`?

Comment (slavek-kucera) [reply parser_library/src/lexing/logical_line.h:149]: range_provider would be a good candidate, but I am fine with keeping it in the db2_logical_line_helper, or just a freestanding function in the db2 preprocessor or range provider source file.

File: parser_library/src/processing/preprocessors/db2_preprocessor.cpp (hunk 15 above)
Comment (jirimosinger) [reply parser_library/src/processing/preprocessors/db2_preprocessor.cpp]: I think it was created only for new documents. Nevertheless, I agree that the `shared_ptr` was a bit of an overkill -> both logical lines now reside in the helper class and are reused even for documents

File: parser_library/src/semantics/range_provider.cpp [hunk 16]
```diff
@@ -169,3 +168,18 @@ range_provider::range_provider()
...
+range text_range(
+    const lexing::logical_line::const_iterator& b, const lexing::logical_line::const_iterator& e, size_t lineno_offset)
+{
+    const auto [bx, by] = b.get_coordinates();
+    position b_pos(by + lineno_offset, bx);
+    if (b == e) // empty range
+        return range(std::move(b_pos));
+    else
+    {
+        const auto [ex, ey] = e.get_coordinates();
```
Comment (slavek-kucera) [inline parser_library/src/semantics/range_provider.cpp:181]: The std::prev(e) should always be here, if your range happens to end at the end of segment, the end iterator points to the next (non-empty) line and the line number will be off.

At the same time my original suggestions did not cover the "empty logical line" case. So, I suggest the past-the-end iterator having coordinates [0,0] by definition i.e.
```
if (m_col == column_iterator()) return {0,0};
```
That way the function is always defined and "by accident" it handles the empty LL case correctly.

**Review by slavek-kucera — APPROVED** [review]

**Pull request merged**